### PR TITLE
[Refactor] Decouple the wallet class

### PR DESCRIPTION
**Pull request by panleone**
## Abstract

The file `wallet.js` is a real mess: There is no actual wallet manager class and it mixes GUI, crypto functions, ledger stuff and masterKeys. This PR begins a process of refactoring the class:

- A Wallet class has been added that at the moment incorporates all the ugly global wallet related functions that were in `wallet.js`. In the future Wallet will manage UTXOs, balance, addresses, masternodes...
- MasterKey class and derivatives have been moved to a new file `masterkey.js`, masterkeys should have the ONLY role of managing internally the generation of new addresses.
- Hence the global variable `masterKey` has been removed in favor of a global variable of type `Wallet` that has `masterKey` as a private field. At the moment I added a getter to `masterKey` (or the refactor would have been to big), in future PRs we should try to avoid exposing masterkey outside the wallet.
- Ledger related functions and variables have been moved to a new file `ledger.js`
- Finally a new file `encoding.js` has been created. It contains the low-level  cryptographic functions that were previously in the wallet file.

---

## Testing
Test that the wallet functionality are still working as expected for example:

- Create wallet
- Import wallet
- Hit 'Refresh Address' in the Dashboard
- Encrypt (+ Change Password)
- Test Send + Receive
- Test Contacts (Add them, Send to them, etc)

---

## Note
As I already said the refactor is not final and I did not want to change too much lines. In future PRs the getter to `masterKey` and the function `getDerivationPath` should both be removed from the `Wallet` class. We should also really try to decouple the logic from the GUI

### CODE REVIEW

**Comment by JSKitty** [comment]
Before I add the "Awaiting Review" label to this PR, the 'Testing' section is going to need to be more detailed (the Quality Control team need to understand *what* to test, too vague and we might not get enough tests by folks).

**Review by JSKitty — COMMENTED** [review]
Code-wise this is looking very clean and is a massive organisation to what we had before, I like it! :+1: :rocket:

I have some minor nits related to the MasterKey and Wallet classes, but many of them are out of scope to the PR (decoupling refactor) so I'll probably work on them in a different PR to keep this small and quick.

To real nits: I'm not sure `cypher.js` quite fits what we're using it for, maybe `encoding.js` would fit better? It is more En/Decoding than cyphering.

I'll get testing in the meantime. :wrench:

**Comment by panleone** [comment]
> To real nits: I'm not sure `cypher.js` quite fits what we're using it for, maybe `encoding.js` would fit better? It is more En/Decoding than cyphering.

totally agree

File: scripts/i18n.js [hunk 1]
```diff
@@ -55,13 +55,14 @@ export function switchTranslation(langName) {
...
-        if (masterKey && cNet) {
+        if (wallet.isLoaded() && cNet) {
             updateEncryptionGUI();
             renderActivityGUI(cNet.arrTxHistory);
         }
         loadAlerts();
         fillAnalyticSelect();
-        if (masterKey) {
+        if (wallet.isLoaded()) {
+            console.log('aszz');
```
[the remark this answers is not in the document]
Comment (panleone) [reply scripts/i18n.js]: fixed lol

**Review by Duddino — APPROVED** [review]
tACK

File: scripts/wallet.js [hunk 2]
```diff
@@ -23,452 +21,145 @@ import {
...
-    return fParseBytes ? Uint8Array.from(bWIF.slice(1, 33)) : true;
-}
-
-// A convenient alias to verifyWIF that returns the raw byte payload
-export function parseWIF(strWIF, skipVerification = false) {
-    return verifyWIF(strWIF, true, skipVerification);
-}
+export const wallet = new Wallet();
 
 /**
```
Comment (Duddino) [inline scripts/wallet.js:164]: Leftover coment

**Review by Duddino — DISMISSED** [review]
Other than the leftover comment, tACK

File: scripts/wallet.js (hunk 2 above)
Comment (panleone) [reply scripts/wallet.js:164]: removed

**Review by JSKitty — APPROVED** [review]
**tACK** e6ddb9e67365c4f1b1a886beacee6028e5016f38

**Review by Liquid369 — APPROVED** [review]
tACK e6ddb9e67365c4f1b1a886beacee6028e5016f38
Ledger is also working again LGTM

**Review by Duddino — APPROVED** [review]
tACK ledger works fine

**Pull request merged**